### ISSUE TITLE
Allow observation extraction from questionnaire response

### ISSUE THREAD

**Issue by maimoonak** [opensrp/fhircore#525]
**Observation based extraction**

**Observations mapping in a questionnaire and response should be converted to fhir Observation resource**

Details here: http://build.fhir.org/ig/HL7/sdc/extraction.html#observation-based-extraction

**Comment by f-odhiambo** [comment]
@maimoonak This sounds like an SDK task. Kindly clarify

**Comment by pld** [comment]
do StructureMaps handle this?

**Comment by maimoonak** [comment]
Yeah. This is SDK task where SDK should provide Observation based extraction. 
Structure Map also handles this. 

There are three ways to handle extraction

http://build.fhir.org/ig/HL7/sdc/extraction.html#designing-questionnaires-to-support-data-extraction
- Observation-based
- Definition-based
- StructureMap-based

**Comment by pld** [comment]
We can use Definition-based if that's simpler

**Comment by pld** [comment]
In which case this would not be an SDK issue, since defining how a QuestionnaireResponse converts to an Observation is a implementation task that goes into a Questionnaire in HAPI and gets synced down to the app (or will when we implement that)

**Comment by maimoonak** [comment]
@pld For the sake of completion of task I have added written a class that does this conversion for us. This can be done with Observation based or Structure Map based only. With definition based problem is each field maps to one property which would be obs.answer only. and we have several observations into one form and trying adding those would make the questionnaires lengthy and difficult to manage. However, for Observation we need a complete mapping of properties of Observation like below (as with current temporary implementation)
this=obs
```
      this.id = getUniqueId()
      this.effective = DateTimeType.now()
      this.code = item.asCodeableConcept() // converts to a codeable concept
      this.status = Observation.ObservationStatus.FINAL
      this.value = answer?.firstOrNull()?.value
      this.subject = patient
```

In short we have found a way around and can wait for SDK implementation or can use StructureMap

**Comment by pld** [comment]
Can we close this issue?

**Comment by maimoonak** [comment]
Nopes. This is not done.

**Comment by pld** [comment]
OK, @f-odhiambo is there any urgency on this? Otherwise, I think we can take it out of the current milestone

**Comment by f-odhiambo** [comment]
Will add some updates here once I have a check-in with Abdul and Raazia who's work is dependent on this

**Comment by f-odhiambo** [comment]
@maimoonak Do we still need this here ?

**Comment by maimoonak** [comment]
For fhir-core we may not need it since we are using structure-map. But its a valid fhir spec feature. So we may want it to be supported  by SDK. 

For fhircore not an issue now.

**Comment by f-odhiambo** [comment]
Closing this for now
